x/poolmanager/taker_fee.go

```diff
@@ -0,0 +1,182 @@
+}
+
+// extractTakerFeeAndDistribute takes in a pool and extracts the taker fee from the pool and sends it to the non native fee pool module account.
+// Its important to note here that in the original swap, the taker fee + spread fee is sent to the pool's address, so this is why we
+// pull directly from the pool and not the user's account.
+func (k Keeper) extractTakerFeeAndDistribute(ctx sdk.Context, tokenIn sdk.Coin, tokenOutDenom string, sender sdk.AccAddress, exactIn bool) (sdk.Coin, error) {
+	nonNativeFeeCollectorForStakingRewardsName := txfeestypes.NonNativeFeeCollectorForStakingRewardsName
+	nonNativeFeeCollectorForCommunityPoolName := txfeestypes.NonNativeFeeCollectorForCommunityPoolName
+	baseDenom := appparams.BaseCoinUnit
+	poolManagerParams := k.GetParams(ctx)
+
+	takerFee, err := k.GetTradingPairTakerFee(ctx, tokenIn.Denom, tokenOutDenom)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	var tokenInAfterTakerFee sdk.Coin
+	var takerFeeCoin sdk.Coin
+	if exactIn {
+		tokenInAfterTakerFee, takerFeeCoin = k.calcTakerFeeExactIn(tokenIn, takerFee)
+	} else {
+		tokenInAfterTakerFee, takerFeeCoin = k.calcTakerFeeExactOut(tokenIn, takerFee)
+	}
+
+	// N.B. We truncate from the community pool calculation, then remove that from the total, and use the remaining for staking rewards.
+	// If we truncate both, these can leave tokens in the users wallet when swapping and exact amount in, which is bad UX.
+
+	// We determine the distributution of the taker fee based on its denom
+	// If the denom is the base denom:
+	takerFeeAmtRemaining := takerFeeCoin.Amount
+	if takerFeeCoin.Denom == baseDenom {
+		// Community Pool:
+		if poolManagerParams.TakerFeeParams.OsmoTakerFeeDistribution.CommunityPool.GT(sdk.ZeroDec()) {
+			// Osmo community pool funds is a direct send
+			osmoTakerFeeToCommunityPoolDec := takerFeeAmtRemaining.ToDec().Mul(poolManagerParams.TakerFeeParams.OsmoTakerFeeDistribution.CommunityPool)
+			osmoTakerFeeToCommunityPoolCoins := sdk.NewCoins(sdk.NewCoin(baseDenom, osmoTakerFeeToCommunityPoolDec.TruncateInt()))
+			err := k.communityPoolKeeper.FundCommunityPool(ctx, osmoTakerFeeToCommunityPoolCoins, sender)
+			if err != nil {
+				return sdk.Coin{}, err
+			}
+			takerFeeAmtRemaining = takerFeeAmtRemaining.Sub(osmoTakerFeeToCommunityPoolCoins.AmountOf(baseDenom))
+		}
+		// Staking Rewards:
+		if poolManagerParams.TakerFeeParams.OsmoTakerFeeDistribution.StakingRewards.GT(sdk.ZeroDec()) {
+			// Osmo staking rewards funds are sent to the non native fee pool module account (even though its native, we want to distribute at the same time as the non native fee tokens)
+			// We could stream these rewards via the fee collector account, but this is decision to be made by governance.
+			osmoTakerFeeToStakingRewardsCoins := sdk.NewCoins(sdk.NewCoin(baseDenom, takerFeeAmtRemaining))
+			err := k.bankKeeper.SendCoinsFromAccountToModule(ctx, sender, nonNativeFeeCollectorForStakingRewardsName, osmoTakerFeeToStakingRewardsCoins)
```

Comment:
Why is this sent to non native fee collector and not staking rewards module account directly if this is already OSMO?

Reply:
Was a discussion I had with Alpin. We could do either, but distributing all at once rather than just trickling osmo rewards while distributing the swapped funds later is minor potential increase of sell pressure. We could absolutely do either way though, and the sell pressure point could be so small that its a non-issue.

Reply:
No strong preferences on this. Just noticed that this additional step wasn't necessary.

Would be good to document that this is a specific design decision we made

Reply:
Also, can we then rename it from "nonNativeFeeCollector" to just "staking fee collector" since it also collects native tokens?

Reply:
Renamed https://github.com/osmosis-labs/osmosis/pull/6034/commits/ecaa3f3145dcddc5e7a98138737fa5696cfb43d2